text/0000-gradual-struct-init.md

```diff
@@ -0,0 +1,739 @@
+straightforward. We simply extend the assignment syntax that already exists.
+
+However, some tweaks in various directions could be made to the RFC:
+
+1. We could limit gradual initialization to top level bindings.
+   This would limit the power and is seemingly an arbitrary restriction
+   that could cause surprises.
+
+1. We could require that a gradually initialized place be definitively
+   initialized before allowing references or moving out of parts.
+   This would mean that we could not write:
+
+   ```rust
+   let foo: Foo;
+   foo.bar = 1;
+   foo.baz = &foo.bar;
+   ```
+
+   That would remove one of the uniquely useful things about the
+   proposed syntax.
+
+3. We could limit this to nominal types or just to types with named fields.
+   While that is possible, it complicates understanding of the type system.
+
+4. We could allow `Drop` types to be gradually initialized.
+   This is discussed in the [unresolved questions][unresolved-questions].
+
+# Prior art
+[prior-art]: #prior-art
+
+None to our knowledge.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1. Should we lift the requirement that each field must be definitively
+   initialized at least once before a gradually initialized place goes
+   out of scope?
+
+   We have imposed this restriction to make sure that adding a new field
+   causes a type error in all cases; if we lifted the restriction,
+   that would no longer hold.
+
+2. Should we allow gradual initialization for types implementing `Drop`?
+
+   While this might make the language more consistent and uniform,
+   it might also become too much magic. In particular, it becomes less
+   clear when destructors will run and understanding when it happens might
+   require tracking "more boolean variables" in your head.
+   For example, if we write:
+
+   ```rust
+   let foo;
+   foo.bar = <expr>;
+   foo.baz = <expr>;
+   foo.quux = <expr>;
+   may_panic();
+
+   // a few lines later...
+   foo.wibble = <expr>;
+   ```
+
+   Then we have to know that `foo` has 4 fields to know that `may_panic()`
+   will not run the destructor of `typeof(foo)`.
+
+   It is however unclear if this would be a problem in practice and
+   it could be solved by privacy.
+
+   If we do however allow `Drop` types to be gradually initialized,
+   then we should also lift the restriction on moving out of `Drop`
+   types to make things consistent.
+
+   If we want to allow `Drop` types, then the restriction in 1.
+   is helpful because there's less to track in your head.
+
+3. If we permit structural records, and if the user writes:
+
+   ```rust
+   let foo;
+   foo.bar = 1;
+   foo.baz = true;
+   foo.qux = "hello";
+   drop(foo);
+   ```
+
+   Should then `foo`, if not otherwise constrained, be inferred to the type?:
+
+   ```rust
+   { bar: i32, baz: bool, qux: &'static str }
+   ```
+
+   This could be quite ergonomic, but it could also be too much magic.
+
```

Comment:
I think initializing arrays by using assignments with constant indices is also one extension that should be discussed somewhere (or at least explicitly punted to the future).

```rust
let arr: [i32; 3];
arr[2] = 42;
arr[0] = 0;
arr[1] = -12;
drop(arr);
```

Reply:
Thanks for the reminder; Manish noted this on Discord but I somehow forgot; I've added a section in "Future possibilities".